### DISCUSSION TITLE
How does direct emitter visibility (from sensor) work in Path integrator ?

### DISCUSSION THREAD

**Discussion by xacond00** [mitsuba-renderer/mitsuba3#1338] [Q&A] [answered]
When there is previous surface interaction `ls.prev_si`, the code makes perfect sense, as the direction sample contains direction towards the previous intersection point, and thus later calling `pdf_direction()` correctly computes cos(theta) term, based on the surface normal and incoming direction.

But how on earth does this code work, when `ls.prev_si` is initialized to `dr::zeros()` for primary rays generated from the sensor ? The direction vector in `ds` now points from the world origin towards the emitter intersection. How can then correct cos(theta) be computed, when the direction should be identical with primary ray's world direction and not `si.p - ref.p` where `ref.p = (0,0,0)` based on the direction sample constructor ? 
 
```C++
if (dr::any_or<true>(si.emitter(scene) != nullptr)) { 
                    // On 0th iteration, the prev_si contains all zeros, meaning ds generates direction towards si.p from world origin ?
                    // But shouldn't the direction be from ray.o towards si.p ???
                    DirectionSample3f ds(scene, si, ls.prev_si);
                    Float em_pdf = 0.f;
                   // Cos theta term is internally computed based on ds.d and si.n
                    if (dr::any_or<true>(!ls.prev_bsdf_delta))
                        em_pdf = scene->pdf_emitter_direction(ls.prev_si, ds, !ls.prev_bsdf_delta);
                    Float mis_bsdf = mis_weight(ls.prev_bsdf_pdf, em_pdf);
                    ls.result = spec_fma(ls.throughput, ds.emitter->eval(si, ls.prev_bsdf_pdf > 0.f) * mis_bsdf, ls.result);
}
```

So I guess my question is, why the `prev_si` isn't initialized with primary ray's origin as the intersection point ? It just doesn't make sense, how we can measure radiance without knowing the prior point ?

**Comment by njroussel** [comment, marked as answer]
Hi @xacond00 

The `prev_bsdf_delta` variable is initialized to `true`, so the entire `em_pdf = scene->pdf_emitter_direction(ls.prev_si, ds, !ls.prev_bsdf_delta);` computation is masked (the second argument of `pdf_emitter_direction` is a mask to indicate which lanes/threads are active). Effectively this means that `em_pdf` is equal to `0` and doesn't depend on the cosine term you mentioned.

Note that `prev_bsdf_pdf` is initialized as `1`, so the `mis_bsdf` is equal to `1` even though `em_pdf` is 0.

**Reply by xacond00** [reply]
Yes, I noticed that too yesterday. Maybe should test more, before bothering anyone here :P
I made a mistake when I was stripping the sample function to compute primary interaction only, and completely missed the negation in there. The `!ls.prev_bsdf_delta` is infact all `false` on the primary interaction, and thus the emitter pdf computation doesn't happen at all. 

Anyways, thanks !
